src/main/java/org/wildfly/security/auth/client/AuthenticationConfiguration.java

```diff
@@ -1321,6 +1325,63 @@ public String toString() {
         }
         return toString;
     }
+    //user callback sanitation
+    private void sanitazeOnMutation(final int what) {
+        try{
+        switch (what) {
+            case SET_PRINCIPAL:
+                // CallbackKind.PRINCIPAL
+                if(this.principal!=null && !this.principal.equals(AnonymousPrincipal.getInstance())){
+                    this.userCallbackKinds.remove(CallbackKind.PRINCIPAL);
+                }
+                break;
+            case SET_CRED_SOURCE:
+                // CallbackKind.CREDENTIAL
+                if(this.credentialSource!=null && !this.credentialSource.equals(IdentityCredentials.NONE)){
+                    this.userCallbackKinds.remove(CallbackKind.CREDENTIAL);
+                }
+                break;
+            case SET_REALM:
+                // CallbackKind.REALM
+                this.userCallbackKinds.remove(CallbackKind.REALM);
+                break;
+            case SET_PARAM_SPECS:
+                // CallbackKind.PARAMETERS
+                this.userCallbackKinds.remove(CallbackKind.PARAMETERS);
+                break;
+            case SET_KEY_MGR_FAC:
+                // CallbackKind.PEER_CREDENTIAL
+                this.userCallbackKinds.remove(CallbackKind.PEER_CREDENTIAL);
+                break;
+            case SET_USER_CB_KINDS:
+                // SANITAZE on above content
+                if (this.principal != null) {
+                    sanitazeOnMutation(SET_PRINCIPAL);
+                }
+
+                if (this.credentialSource != null) {
```

Comment:
It is already checked inside of `sanitazeOnMutation(SET_CRED_SOURCE)` - this condition is only small optimization - maybe would be better to remove it.